File: grey-haven-plugins/hooks/hooks/js/work-completion-assistant.js
#!/usr/bin/env node

/**
 * Work Completion Assistant Hook for JavaScript/TypeScript
 * =========================================================
 * Type: Stop
 * Description: Ensures work is complete before allowing Claude to stop
 *
 * This hook validates that all work has been properly completed, including:
 * - Tests are passing
 * - No unresolved TODOs
 * - Code is formatted
 * - Documentation is updated
 * - Changes are committed (optionally)
 */

const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

class WorkCompletionValidator {
    constructor(projectDir) {
        this.projectDir = projectDir;
        this.issues = [];
        this.warnings = [];
    }

    checkUncommittedChanges() {
        try {
            const status = execSync('git status --porcelain', {
                encoding: 'utf8',
                cwd: this.projectDir
            }).trim();

            if (status) {
                const changedFiles = status.split('\n').length;
                this.warnings.push(`You have ${changedFiles} uncommitted changes`);
                return false;
            }
            return true;
        } catch {
            return true; // Not a git repo, skip
        }
    }

    checkTodosInChangedFiles() {
        try {
            // Get changed files
            const diffFiles = execSync('git diff --name-only HEAD', {
                encoding: 'utf8',
                cwd: this.projectDir
            }).trim().split('\n').filter(Boolean);

            const stagedFiles = execSync('git diff --cached --name-only', {
                encoding: 'utf8',
                cwd: this.projectDir
            }).trim().split('\n').filter(Boolean);

            const allFiles = [...new Set([...diffFiles, ...stagedFiles])];
            const filesWithTodos = [];

            for (const file of allFiles) {
                const filePath = path.join(this.projectDir, file);
                if (fs.existsSync(filePath)) {
                    try {
                        const content = fs.readFileSync(filePath, 'utf8');
                        if (/TODO|FIXME|XXX|HACK/.test(content)) {
                            filesWithTodos.push(file);
                        }
                    } catch {}
                }
            }

            if (filesWithTodos.length) {
                this.issues.push(`Unresolved TODOs in: ${filesWithTodos.slice(0, 3).join(', ')}`);
                return false;
            }
            return true;
        } catch {
            return true;
        }
    }

    checkTests() {
        const packageJsonPath = path.join(this.projectDir, 'package.json');
        
        if (!fs.existsSync(packageJsonPath)) {
            return true; // No package.json, skip
        }

        try {
            const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, 'utf8'));
            const scripts = packageJson.scripts || {};
            
            // Check for test scripts
            const testCommands = [];
            if (scripts.test) {
                testCommands.push('test');
            }
            if (scripts['test:unit']) {
                testCommands.push('test:unit');
            }
            
            if (testCommands.length) {
                // Check if using bun or npm
                const runner = fs.existsSync(path.join(this.projectDir, 'bun.lockb')) ? 'bun' : 'npm';
                this.warnings.push(`Remember to run tests: ${runner} run ${testCommands[0]}`);
            }
            
            // Check for test configuration files
            const testConfigs = [
                'jest.config.js',
                'jest.config.ts',
                'vitest.config.js',
                'vitest.config.ts',
                '.mocharc.json',
                'playwright.config.js',
                'cypress.config.js'
            ];
            
            for (const config of testConfigs) {
                if (fs.existsSync(path.join(this.projectDir, config))) {
                    const testRunner = config.split('.')[0].split('rc')[0];
                    this.warnings.push(`${testRunner} configuration found - ensure tests pass`);
                    break;
                }
            }
            
            return true; // Don't block on tests
        } catch {
            return true;
        }
    }

    checkLinting() {
        // Check for linting configs
        const lintConfigs = [
            'eslint.config.js',
            '.eslintrc.js',
            '.eslintrc.json',
            'biome.json',
            '.prettierrc',
            'prettier.config.js'
        ];

        for (const config of lintConfigs) {
            if (fs.existsSync(path.join(this.projectDir, config))) {
                const linter = config.includes('eslint') ? 'ESLint' : 
                              config.includes('prettier') ? 'Prettier' : 
                              config.includes('biome') ? 'Biome' : 'Linter';
                this.warnings.push(`Remember to run ${linter}`);
                break;
            }
        }

        return true; // Don't block on linting
    }

    checkDocumentation() {
        try {
            // Check if code files were changed
            const changedFiles = execSync('git diff --name-only HEAD', {
                encoding: 'utf8',
                cwd: this.projectDir
            }).trim().split('\n').filter(Boolean);

            const codeFilesChanged = changedFiles.some(f => 
                /\.(js|jsx|ts|tsx|mjs|cjs)$/.test(f)
            );

            const docsChanged = changedFiles.some(f => 
                /\.(md|mdx)$/.test(f) || f.includes('docs/')
            );

            if (codeFilesChanged && !docsChanged) {
                this.warnings.push('Code changed but documentation not updated');
            }

            // Check for JSDoc in TypeScript/JavaScript files
            if (codeFilesChanged) {
                const undocumentedExports = [];
                
                for (const file of changedFiles.filter(f => /\.(js|ts|jsx|tsx)$/.test(f)).slice(0, 3)) {
                    const filePath = path.join(this.projectDir, file);
                    if (fs.existsSync(filePath)) {
                        try {
                            const content = fs.readFileSync(filePath, 'utf8');
                            
                            // Check for exported functions without JSDoc
                            const exportPattern = /export\s+(async\s+)?function\s+(\w+)/g;
                            let match;
                            while ((match = exportPattern.exec(content)) !== null) {
                                const funcName = match[2];
                                const beforeFunc = content.substring(Math.max(0, match.index - 200), match.index);
                                if (!beforeFunc.includes('/**')) {
                                    undocumentedExports.push(`${file}:${funcName}`);
                                }
                            }
                        } catch {}
                    }
                }
                
                if (undocumentedExports.length) {
                    this.warnings.push(`Undocumented exports: ${undocumentedExports.slice(0, 3).join(', ')}`);
                }
            }

            return true; // Don't block on documentation
        } catch {
            return true;
        }
    }

    checkBranches() {
        try {
            const branch = execSync('git branch --show-current', {
                encoding: 'utf8',
                cwd: this.projectDir
            }).trim();

            if (['main', 'master', 'production'].includes(branch)) {
                this.warnings.push(`Working directly on ${branch} branch`);
            }

            return true; // Don't block on branch
        } catch {
            return true;
        }
    }

    checkTypeScript() {
        // Check for TypeScript errors
        const tsconfigPath = path.join(this.projectDir, 'tsconfig.json');
        
        if (fs.existsSync(tsconfigPath)) {
            this.warnings.push('Remember to check TypeScript compilation: npx tsc --noEmit');
            
            // Check for any .ts files with @ts-ignore or @ts-expect-error
            try {
                const tsFiles = execSync('find . -name "*.ts" -o -name "*.tsx" 2>/dev/null | head -20', {
                    encoding: 'utf8',
                    cwd: this.projectDir
                }).trim().split('\n').filter(Boolean);
                
                let ignoreCount = 0;
                for (const file of tsFiles) {
                    const filePath = path.join(this.projectDir, file);
                    if (fs.existsSync(filePath)) {
                        try {
                            const content = fs.readFileSync(filePath, 'utf8');
                            const matches = content.match(/@ts-ignore|@ts-expect-error/g);
                            if (matches) {
                                ignoreCount += matches.length;
                            }
                        } catch {}
                    }
                }
                
                if (ignoreCount > 0) {
                    this.warnings.push(`Found ${ignoreCount} TypeScript ignore comments`);
                }
            } catch {}
        }
        
        return true;
    }

    validate() {
        // Run all checks
        this.checkUncommittedChanges();
        this.checkTodosInChangedFiles();
        this.checkTests();
        this.checkLinting();
        this.checkDocumentation();
        this.checkBranches();
        this.checkTypeScript();

        // Determine if we should block
        const shouldBlock = this.issues.length > 0;

        if (shouldBlock) {
            let reason = 'Work incomplete:\n';
            reason += this.issues.map(issue => `❌ ${issue}`).join('\n');
            
            if (this.warnings.length) {
                reason += '\n\nWarnings:\n';
                reason += this.warnings.map(warning => `⚠️  ${warning}`).join('\n');
            }

            return {
                decision: 'block',
                reason: reason + '\n\nPlease address these items before stopping.'
            };
        } else if (this.warnings.length) {
            // Just show warnings, don't block
            const summary = 'Work appears complete. Reminders:\n' +
                          this.warnings.map(warning => `⚠️  ${warning}`).join('\n');
            
            console.error(summary); // Use stderr for output
            return {};
        } else {
            console.error('✅ All checks passed - work appears complete!');
            return {};
        }
    }
}

// Main execution
let inputData = '';
process.stdin.on('data', chunk => inputData += chunk);
process.stdin.on('end', () => {
    try {
        const data = JSON.parse(inputData);
        
        // Check if we're already in a stop hook loop
        if (data.stop_hook_active) {
            process.exit(0); // Don't create infinite loops
        }
        
        const projectDir = data.cwd || process.cwd();
        
        const validator = new WorkCompletionValidator(projectDir);
        const result = validator.validate();
        
        if (result.decision) {
            console.log(JSON.stringify(result));
        }
        
        process.exit(0);
    } catch (error) {
        // Don't block on errors
        process.exit(0);
    }
});

File: .claude/hooks/hooks/python/work-completion-assistant.py
#!/usr/bin/env -S uv run --quiet --script
# /// script
# dependencies = ["gitpython"]
# ///
"""
Work Completion Assistant Hook
==============================
Type: Stop
Description: Ensures work is complete before allowing Claude to stop

This hook validates that all work has been properly completed, including:
- Tests are passing
- No unresolved TODOs
- Code is formatted
- Documentation is updated
- Changes are committed (optionally)
"""

import json
import sys
import subprocess
from pathlib import Path
from typing import List, Dict, Optional


class WorkCompletionValidator:
    def __init__(self, project_dir: str):
        self.project_dir = Path(project_dir)
        self.issues = []
        self.warnings = []
        
    def check_uncommitted_changes(self) -> bool:
        """Check for uncommitted changes"""
        try:
            result = subprocess.run(
                ['git', 'status', '--porcelain'],
                capture_output=True, text=True, cwd=self.project_dir
            )
            if result.stdout.strip():
                changed_files = len(result.stdout.strip().split('\n'))
                self.warnings.append(f"You have {changed_files} uncommitted changes")
                return False
            return True
        except:
            return True  # If not a git repo, skip this check
    
    def check_todos_in_changed_files(self) -> bool:
        """Check for TODOs in recently changed files"""
        try:
            # Get list of changed files
            result = subprocess.run(
                ['git', 'diff', '--name-only', 'HEAD'],
                capture_output=True, text=True, cwd=self.project_dir
            )
            changed_files = result.stdout.strip().split('\n') if result.stdout.strip() else []
            
            # Also check staged files
            result = subprocess.run(
                ['git', 'diff', '--cached', '--name-only'],
                capture_output=True, text=True, cwd=self.project_dir
            )
            staged_files = result.stdout.strip().split('\n') if result.stdout.strip() else []
            
            all_files = set(changed_files + staged_files)
            files_with_todos = []
            
            for file in all_files:
                if file and Path(self.project_dir / file).exists():
                    try:
                        with open(self.project_dir / file, 'r') as f:
                            content = f.read()
                            if 'TODO' in content or 'FIXME' in content or 'XXX' in content:
                                files_with_todos.append(file)
                    except:
                        pass
            
            if files_with_todos:
                self.issues.append(f"Unresolved TODOs in: {', '.join(files_with_todos[:3])}")
                return False
            return True
        except:
            return True
    
    def check_tests(self) -> bool:
        """Run tests if test command is available"""
        test_commands = []
        
        # Check for different test runners
        if (self.project_dir / 'package.json').exists():
            try:
                with open(self.project_dir / 'package.json', 'r') as f:
                    package_json = json.load(f)
                    scripts = package_json.get('scripts', {})
                    if 'test' in scripts:
                        test_commands.append('npm test')
                    elif 'test:unit' in scripts:
                        test_commands.append('npm run test:unit')
            except:
                pass
        
        if (self.project_dir / 'pyproject.toml').exists():
            if subprocess.run(['which', 'pytest'], capture_output=True).returncode == 0:
                test_commands.append('pytest --co -q')  # Just collect tests, don't run
            elif subprocess.run(['which', 'python'], capture_output=True).returncode == 0:
                test_commands.append('python -m pytest --co -q')
        
        if (self.project_dir / 'go.mod').exists():
            test_commands.append('go test ./... -run=^$ -count=1')  # Dry run
        
        if (self.project_dir / 'Cargo.toml').exists():
            test_commands.append('cargo test --no-run')  # Compile but don't run
        
        # For now, just warn about tests, don't block
        if test_commands:
            self.warnings.append(f"Remember to run tests: {test_commands[0]}")
        
        return True  # Don't block on tests
    
    def check_linting(self) -> bool:
        """Check if linting would pass"""
        lint_commands = []
        
        # Detect linters
        if (self.project_dir / 'pyproject.toml').exists():
            if subprocess.run(['which', 'ruff'], capture_output=True).returncode == 0:
                result = subprocess.run(
                    ['ruff', 'check', '--quiet'],
                    capture_output=True, cwd=self.project_dir
                )
                if result.returncode != 0:
                    self.warnings.append("Code has linting issues (ruff)")
        
        if (self.project_dir / 'eslint.config.js').exists() or \
           (self.project_dir / '.eslintrc.js').exists():
            # Just warn, don't run eslint as it might be slow
            self.warnings.append("Remember to run eslint")
        
        return True  # Don't block on linting
    
    def check_documentation(self) -> bool:
        """Check if documentation might need updating"""
        try:
            # Check if any code files were changed
            result = subprocess.run(
                ['git', 'diff', '--name-only', 'HEAD'],
                capture_output=True, text=True, cwd=self.project_dir
            )
            changed_files = result.stdout.strip().split('\n') if result.stdout.strip() else []
            
            code_files_changed = any(
                f.endswith(('.py', '.js', '.ts', '.go', '.rs', '.java'))
                for f in changed_files
            )
            
            docs_changed = any(
                f.endswith(('.md', '.mdx', '.rst', '.txt')) or 'docs/' in f
                for f in changed_files
            )
            
            if code_files_changed and not docs_changed:
                self.warnings.append("Code changed but documentation not updated")
            
            return True  # Don't block on documentation
        except:
            return True
    
    def check_branches(self) -> bool:
        """Check if working on correct branch"""
        try:
            result = subprocess.run(
                ['git', 'branch', '--show-current'],
                capture_output=True, text=True, cwd=self.project_dir
            )
            branch = result.stdout.strip()
            
            if branch in ['main', 'master', 'production']:
                self.warnings.append(f"Working directly on {branch} branch")
            
            return True  # Don't block on branch
        except:
            return True
    
    def validate(self) -> Dict:
        """Run all validations and return results"""
        # Run checks
        self.check_uncommitted_changes()
        self.check_todos_in_changed_files()
        self.check_tests()
        self.check_linting()
        self.check_documentation()
        self.check_branches()
        
        # Determine if we should block
        should_block = len(self.issues) > 0
        
        if should_block:
            reason = "Work incomplete:\n"
            reason += "\n".join(f"❌ {issue}" for issue in self.issues)
            if self.warnings:
                reason += "\n\nWarnings:\n"
                reason += "\n".join(f"⚠️  {warning}" for warning in self.warnings)
            
            return {
                "decision": "block",
                "reason": reason + "\n\nPlease address these items before stopping."
            }
        elif self.warnings:
            # Just show warnings, don't block
            summary = "Work appears complete. Reminders:\n"
            summary += "\n".join(f"⚠️  {warning}" for warning in self.warnings)
            
            # Print to stdout for transcript mode
            print(summary)
            
            return {}  # Don't block
        else:
            print("✅ All checks passed - work appears complete!")
            return {}


def main():
    try:
        input_data = json.load(sys.stdin)
    except json.JSONDecodeError:
        sys.exit(1)
    
    # Check if we're already in a stop hook loop
    if input_data.get('stop_hook_active', False):
        # Don't create infinite loops
        sys.exit(0)
    
    project_dir = input_data.get('cwd', '.')
    
    validator = WorkCompletionValidator(project_dir)
    result = validator.validate()
    
    if result:
        print(json.dumps(result))
    
    sys.exit(0)


if __name__ == "__main__":
    main()